scipy/spatial/distance.py

```diff
@@ -1150,53 +1136,64 @@ def sokalsneath(u, v):
-                  'minkowski', 'rogerstanimoto', 'russellrao', 'seuclidean',
-                  'sokalmichener', 'sokalsneath', 'sqeuclidean', 'yule', 'wminkowski']
-
-_TEST_METRICS = {'test_' + name: eval(name) for name in _METRICS_NAMES}
+_convert_to_double = partial(_convert_to_type, out_type=np.double)
+_convert_to_bool = partial(_convert_to_type, out_type=bool)
+
+# adding python-only wrappers to _distance_wrap module
+_distance_wrap.pdist_correlation_double_wrap = _correlation_pdist_wrap
+_distance_wrap.cdist_correlation_double_wrap = _correlation_cdist_wrap
+
+# Registry of implemented metrics:
+# Structure:
+#{
+# metric_name :                         # must be equal to python metric name
+# {'aka' : [list of aliases],
+#
+#  'types': [list of supported types],  # X (pdist) and XA (cdist) are used to choose
+#                                       # the type. if there is no match the first
+#                                       # type is used. Default double
+#
+#  'validator': f(X, m, n, **kwargs)    # function that check kwargs and computes
+#                                       # default values.
+#}
+
+_METRICS = {
+    'braycurtis': {'aka': ['braycurtis']},
+    'canberra': {'aka': ['canberra']},
+    'chebyshev': {'aka': ['chebychev', 'chebyshev', 'cheby', 'cheb', 'ch']},
+    'cityblock': {'aka': ['cityblock', 'cblock', 'cb', 'c']},
+    'correlation': {'aka': ['correlation', 'co']},
+    'cosine': {'aka': ['cosine', 'cos']},
+    'dice': {'aka': ['dice'], 'types': ['bool']},
+    'euclidean': {'aka': ['euclidean', 'euclid', 'eu', 'e']},
+    'hamming': {'aka': ['matching', 'hamming', 'hamm', 'ha', 'h'],
+                'types': ['double','bool']},
+    'jaccard': {'aka': ['jaccard', 'jacc', 'ja', 'j'], 'types': ['double','bool']},
+    'kulsinski': {'aka': ['kulsinski'], 'types': ['bool']},
+    'mahalanobis': {'aka': ['mahalanobis', 'mahal', 'mah'],
+                    'validator': _validate_mahalanobis_kwargs},
+    'minkowski': {'aka':['minkowski', 'mi', 'm'],
+                  'validator': _validate_minkowski_kwargs},
+    'rogerstanimoto': {'aka': ['rogerstanimoto'], 'types': ['bool']},
+    'russellrao': {'aka': ['russellrao'], 'types': ['bool']},
+    'seuclidean': {'aka': ['seuclidean', 'se', 's'],
+                   'validator': _validate_seuclidean_kwargs},
+    'sokalmichener': {'aka': ['sokalmichener'], 'types': ['bool']},
+    'sokalsneath': {'aka': ['sokalsneath'], 'types': ['bool']},
+    'sqeuclidean': {'aka': ['sqeuclidean', 'sqe', 'sqeuclid']},
+    'wminkowski': {'aka': ['wminkowski', 'wmi', 'wm', 'wpnorm'],
+                   'validator': _validate_wminkowski_kwargs},
+    'yule': {'aka': ['yule'], 'types': ['bool']},
+    }
+
+
+_METRIC_ALIAS = dict((alias, name)
+                     for name, info in _METRICS.items()
+                     for alias in info['aka'])
+
+_METRICS_NAMES = list(_METRICS.keys())
+
+_TEST_METRICS = {'test_' + name: eval(name) for name in _METRICS.keys()}
```

Comment:
I still don't like the use of `eval`, especially when it's just being used for name lookup.